x/dex/contract/abci.go

```diff
@@ -154,22 +194,36 @@ func orderMatchingRunnable(ctx context.Context, sdkContext sdk.Context, env *env
 	if !contractInfo.NeedOrderMatching {
 		return
 	}
+	parentSdkContext := sdkContext
 	sdkContext = decorateContextForContract(sdkContext, contractInfo)
 	sdkContext.Logger().Info(fmt.Sprintf("End block for %s", contractInfo.ContractAddr))
-	if orderResultsMap, settlements, err := HandleExecutionForContract(ctx, sdkContext, contractInfo, keeper, tracer); err != nil {
+	pairs, pairFound := env.registeredPairs.Load(contractInfo.ContractAddr)
+	orderBooks, found := env.orderBooks.Load(contractInfo.ContractAddr)
+	if !pairFound || !found {
+		sdkContext.Logger().Error(fmt.Sprintf("No pair or order book for %s", contractInfo.ContractAddr))
+		env.failedContractAddresses.Add(contractInfo.ContractAddr)
+	} else if orderResultsMap, settlements, err := HandleExecutionForContract(ctx, sdkContext, contractInfo, keeper, pairs, orderBooks, tracer); err != nil {
 		sdkContext.Logger().Error(fmt.Sprintf("Error for EndBlock of %s", contractInfo.ContractAddr))
 		env.failedContractAddresses.Add(contractInfo.ContractAddr)
 	} else {
 		for account, orderResults := range orderResultsMap {
 			// only add to finalize message for contract addresses
 			if msg, ok := env.finalizeBlockMessages.Load(account); ok {
+				// ordering of `AddContractResult` among multiple orderMatchingRunnable instances doesn't matter
+				// since it's not persisted as state, and it's only used for invoking registered contracts'
+				// FinalizeBlock sudo endpoints, whose state updates are gated by whitelist stores anyway.
 				env.finalizeMsgMutex.Lock()
 				msg.AddContractResult(orderResults)
 				env.finalizeMsgMutex.Unlock()
 			}
 		}
 		env.settlementsByContract.Store(contractInfo.ContractAddr, settlements)
 	}
+
+	// ordering of events doesn't matter since events aren't part of consensus
+	env.eventManagerMutex.Lock()
+	parentSdkContext.EventManager().EmitEvents(sdkContext.EventManager().Events())
```

Comment:
nit - is it possible for this to throw error? should we use defer for the mutex?

Reply:
yeah makes sense